fold/fold.go

```diff
@@ -0,0 +1,895 @@
+package fold
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/TimothyStiles/poly/transform"
+)
+
+// Fold the DNA sequence and return the lowest free energy score.
+//
+// Based on the approach described in:
+// Zuker and Stiegler, 1981
+// https://www.ncbi.nlm.nih.gov/pmc/articles/PMC326673/pdf/nar00394-0137.pdf
+//
+// If the sequence is 50 or more bp long, "isolated" matching bp
+// are ignored in V(start,end). This is based on an approach described in:
+// Mathews, Sabina, Zuker and Turner, 1999
+// https://www.ncbi.nlm.nih.gov/pubmed/10329189
+// Args:
+//
+//	seq: The sequence to Fold
+//	temp: The temperature the Fold takes place in, in Celsius
+//
+// Returns a slice of NucleicAcidStructure with the energy and description,
+// i.e. stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) ([]NucleicAcidStructure, error) {
+	foldContext, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error creating folding context: %w", err)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return Traceback(0, len(seq)-1, foldContext), nil
+}
+
+// MinimumFreeEnergy folds the sequence and return just the delta G of the structure
+// Args:
+//
+//	seq: The sequence to fold
+//	temp: The temperature to fold at
+//
+// Returns: the minimum free energy of the folded sequence
+func MinimumFreeEnergy(seq string, temp float64) (float64, error) {
+	NucleicAcidStructures, err := Fold(seq, temp)
+	if err != nil {
+		return 0, fmt.Errorf("error folding: %w", err)
+	}
+	summedEnergy := 0.0
+	for _, structure := range NucleicAcidStructures {
+		summedEnergy += structure.Energy
+	}
+	return RoundFloat(summedEnergy, 2), nil
+}
+
+// Get the dot bracket notation for a secondary structure.
+// Args:
+//
+//	NucleicAcidStructures: A list of NucleicAcidStructure, usually from the fold function
+//
+// # Returns the dot-bracket notation of the secondary structure
+//
+// Dot-bracket notation, consisting in a balanced parentheses string composed
+// by a three-character alphabet {.,(,)}, that can be unambiguously converted
+// in the RNA secondary structure. See example_test.go for a small example.
+func DotBracket(NucleicAcidStructures []NucleicAcidStructure) string {
+	maxj := 0
```

Comment:
what is maxj?

Reply:
Seems the be the max of the `End` of the internal subsequences.